PROTOCOL.md

```diff
@@ -427,63 +427,204 @@ Field    | Type | Description
+27 | [StreamMessage](#streammessage)
 28 | [GroupKeyRequest](#groupkeyrequest)
 29 | [GroupKeyResponse](#groupkeyresponse)
 30 | [GroupKeyReset](#groupkeyreset)
 31 | [GroupKeyErrorResponse](#groupkeyerrorresponse)
+32 | [GroupKeyRotate](#groupkeyrotate)
 
-#### `signatureType`
+#### `contentType`
 
-`signatureType` | Name | Description | Signature payload fields to be concatenated in order
--------------- | ---- |------------ | -----------------------
-0 | `NONE` | No signature. signature field is empty in this case. | None.
-1 | `ETH_LEGACY` | Ethereum signature produced by old clients (Message Layer version 29). The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `publisherId`, `content`
-2 | `ETH` | Ethereum signature produced by current clients (since Message Layer version 30). The signature field is encoded as a hex string. | all the `msgId` fields, (`streamId`, `streamPartition`, `timestamp`, `sequenceNumber`, `publisherId`, `msgChainId`), `prevMsgRef`, `content`
+Describes how the content (post-decryption in case it's encrypted) should be parsed.
+
+`contentType` | Description
+------------- | -----------
+0 | JSON
+
+Other content types, including binary types, will be defined in the future.
 
 #### `encryptionType`
 
+For each non-zero `encryptionType`, the `contentType` is a hex-encoded string.
+
 `encryptionType` | Description
 -------------- | --------
 0 | Unencrypted, plaintext message.
-RSA | Content is asymetrically encrypted using RSA.
-AES | Content is symmetrically encrypted using AES.
-NEW_KEY_AND_AES | The content contains a concatenation of a new AES key and the content, encrypted with the old key.
+1 | Content is asymetrically encrypted using RSA.
+2 | Content is symmetrically encrypted using AES.
+
+#### `signatureType`
+
+`signatureType` | Name | Description | Signature payload fields to be concatenated in order
+-------------- | ---- |------------ | -----------------------
+0 | `NONE` | No signature. The `signature` field is null in this case. | None.
+1 | `ETH_LEGACY` | Ethereum signature produced by old clients. The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `publisherId`, `content`
+2 | `ETH` | Ethereum signature produced by current clients (since Stream Layer version 30). The signature field is encoded as a hex string. | `streamId`, `streamPartition`, `timestamp`, `sequenceNumber`, `publisherId`, `msgChainId`, `prevMsgRef.timestamp`, `prevMsgRef.sequenceNumber`, `content`
+
+### StreamMessage
+
+Contains an arbitrary, application-specific `content` payload, i.e. the `content` of this message should be passed to the subscribing application's message handler. This is the message type that applications interact with via publishing and subscribing to messages.
+
+Example (unencrypted JSON message):
+```
+[32, [...msgIdFields], [...msgRefFields], 27, 0, 0, "{\"content\":42}", 2, "0x29c057786Fa..."]
```

Comment:
The JSON message doesn't need to be inside the `"content":` key, right? i.e. example of "content" shouldn't include a JSON object just containing a "content" key, lest people think this is part of the protocol. It isn't, is it?

Reply:
Clarified